File: frontend/src/languages.ts
import type { Language } from "./services/api";

export const FALLBACK_LANGUAGES: Language[] = [
  { code: "eng_Latn", label: "English", provider: "hf", common: true },
  { code: "pai_Latn", label: "Paite", provider: "hf", common: true },
  { code: "lus_Latn", label: "Mizo", provider: "google", common: true },
  { code: "mni_Beng", label: "Meitei", provider: "google", common: true },
  { code: "mya_Mymr", label: "Burmese", provider: "google", common: true },
  { code: "hin_Deva", label: "Hindi", provider: "google", common: true },
  { code: "asm_Beng", label: "Assamese", provider: "google", common: false },
  { code: "ben_Beng", label: "Bengali", provider: "google", common: false },
  { code: "brx_Deva", label: "Bodo", provider: "google", common: false },
  { code: "zho_Hans", label: "Chinese", provider: "google", common: false },
  { code: "fra_Latn", label: "French", provider: "google", common: false },
  { code: "ind_Latn", label: "Indonesian", provider: "google", common: false },
  { code: "kha_Latn", label: "Khasi", provider: "google", common: false },
  { code: "trp_Latn", label: "Kokborok", provider: "google", common: false },
  { code: "rus_Cyrl", label: "Russian", provider: "google", common: false },
  { code: "spa_Latn", label: "Spanish", provider: "google", common: false },
];

export function groupLanguages(languages: Language[]) {
  const common = languages.filter((lang) => lang.common);
  const more = languages
    .filter((lang) => !lang.common)
    .sort((a, b) => a.label.localeCompare(b.label));
  return { common, more };
}

File: backend/languages.py
"""Language registry: Paite via HF NLLB, everything else via Google Translate."""

ENGLISH = "eng_Latn"
PAITE = "pai_Latn"

LANGUAGES: dict[str, dict] = {
    ENGLISH: {"label": "English", "google": "en", "provider": "hf"},
    PAITE: {"label": "Paite", "google": None, "provider": "hf"},
    "lus_Latn": {"label": "Mizo", "google": "lus", "provider": "google"},
    "mni_Beng": {"label": "Meitei", "google": "mni", "provider": "google"},
    "mya_Mymr": {"label": "Burmese", "google": "my", "provider": "google"},
    "hin_Deva": {"label": "Hindi", "google": "hi", "provider": "google"},
}

HF_LANGUAGES = {ENGLISH, PAITE}


def language_codes() -> list[str]:
    return list(LANGUAGES.keys())


def language_label(code: str) -> str:
    entry = LANGUAGES.get(code)
    return entry["label"] if entry else code


def google_code(code: str) -> str | None:
    entry = LANGUAGES.get(code)
    if not entry:
        return None
    return entry.get("google")


def languages_for_api() -> list[dict]:
    return [
        {
            "code": code,
            "label": meta["label"],
            "provider": meta["provider"],
        }
        for code, meta in LANGUAGES.items()
    ]


def uses_google(src: str, tgt: str) -> bool:
    if src == tgt:
        return False
    if src in HF_LANGUAGES and tgt in HF_LANGUAGES:
        return False
    return True


def resolve_route(src: str, tgt: str) -> str:
    """Returns: identity | hf | google | to_paite | from_paite"""
    if src == tgt:
        return "identity"
    if {src, tgt} == HF_LANGUAGES:
        return "hf"
    if tgt == PAITE:
        return "to_paite"
    if src == PAITE:
        return "from_paite"
    return "google"
